### DISCUSSION TITLE
DAO Roadmap

### DISCUSSION THREAD

**Discussion by lzrscg** [lzrscg/concorde#8] [Roadmap]
> _This is a sub-roadmap. See the [main roadmap](https://github.com/lzrscg/concorde/discussions/11)._

# Objective
Any individual or team should be able to contribute to Concorde's UI and business logic. This not only includes implementing features (like in the case of bug bounties), but also _planning_ new features and strategies.

Contributions should be organized such that one contribution does not interfere with another. If marginal productivity **does not** decrease as a new contributor is added, then Concorde will be able to scale massively.

# Tenets
## Coordinating with large groups scales poorly
One of the great innovations of Bitcoin is that mining is permissionless and therefore allows many participants to secure the network without requiring additional overhead to achieve consensus.

This is unlike most other systems that exist, in which adding people or resources has diminishing returns. In the case of software companies, adding more programmers increases the overhead of collaboration. This is known as [Brooks' Law](https://twitter.com/lzrscg/status/1479233110214864896).
<img width="500" src="https://user-images.githubusercontent.com/84213303/150044287-8d50c580-1e0e-4d57-9208-ea2694407e50.jpg">

## Fork-Based Contributions
Forking is permissionless. Any open source codebase can be forked without coordination between the original author(s) and the forker. This enables any person to make any change to the app they choose without collaboration overhead.
<img width="422" src="https://user-images.githubusercontent.com/84213303/150045211-5899ad27-e2be-4ed9-b71e-c679a95cefa5.png">

However, coordination is needed when merging back to the main (canonical) branch. This is because there needs to be a global consensus among all developers as to which changes go into that branch. 
<img width="617" src="https://user-images.githubusercontent.com/84213303/150045490-fa9b8414-de71-4939-be64-43138c51266c.png">

Therefore, to achieve scalability, there must be no canonical branch. Instead of one main branch that represents the singular true version of the app, every branch must be treated like a viable version. If these different versions of the app are bundled _properly_ within the same UX, all of their changes can be aggregated into one product.
<img width="691" src="https://user-images.githubusercontent.com/84213303/150045938-26c0fcd6-8717-4c90-86a7-0a6ec8ac921d.png">

## Central planning doesn't work
Reaching consensus is hard. Think about how hard it is for Congress to make decisions (this is the latest data at the time of posting). They are trying to reach consensus at a huge scale—and clearly people are unhappy with their ability to get the right things done.
<img width="500" src="https://user-images.githubusercontent.com/84213303/150047687-ed4e6bad-1435-4304-8db3-1a8cee60cbeb.png">

As mentioned in the previous tenet, having a main/canonical branch of code isn't a scalable way to develop software because contributors need to come to consensus on what changes to merge into it.

An easy way to get around the need to achieve consensus is to engage in central planning. If a "decision maker" chooses what changes to put in the main branch, then nobody has to spend any time reaching consensus. This doesn't work for a few reasons.
1. Although consensus is not holding back progress, this causes a new bottleneck to arise. The speed at which the decision maker works holds back progress on the app. How fast can they delegate tasks, analyze data, and plan new features? If you try adding more decision makers to increase capacity, you come up with a consensus problem again.
<img width="500" src="https://user-images.githubusercontent.com/84213303/150049347-9b624a4b-4e0a-4577-b982-bbd71f2963ad.png">

2. We can't be sure that the decision maker will even make good decisions.
3. Central planning has never worked in a political/economic context. There is no reason to believe it is a good system for product development at scale.

## Markets and prices
The market is a decentralized system that generates prices. Prices quantify the relative value between goods and services. For example, if an ounce gold is worth $1000 and an ounce of silver is worth $20, then it would wise to trade an ounce of silver for an ounce of gold, if given the chance.

If a central planner tried to create a ranking list in order to comparatively value different resources, that would be an exhausting
endeavor—increasing combinatorially in difficulty as the number of resources increases. However, the price system emerges effortlessly because it is generated in a decentralized way through the market system.

Think of every individual in the market as a node in a giant neural super computer. These nodes each offer their own computational capacity to the network. Knowledge about the world is also distributed across this network. Each node has a subset of total knowledge and it is recorded in their own local storage. **The market is an efficient algorithm for using this distributed compute power to aggregate all human knowledge in order to find an optimal allocation of resources.** In fact, there is no other known way to utilize all human knowledge than to create a market-like system. 

Through this system, we have had unbounded compounding growth.
![gdp-world](https://user-images.githubusercontent.com/84213303/150182499-18c1a246-7b0f-48dc-8b6a-47a4c4091d61.jpg)

## Discounted Future Cashflow
Discounted future cashflow is a valuation model which dictates that an equity should be valued at the discount rate of all future cashflow.
<img width="378" src="https://user-images.githubusercontent.com/84213303/150183055-ec42fba2-dcb5-4c76-b78e-dd3bb24dc9bc.png">

In layman's terms, this means that if a company will earn $1B over its lifetime, then it is worth approximately $1B. However, since money in the future is worth less than money today (see [net present value](https://en.wikipedia.org/wiki/Net_present_value)), then it will actually be worth less than $1B. So this formula also adjusts for that.

This is a **theoretically sound** way of valuing companies, however, in practice it is hard to know what the future cashflow and discount rate will be.

# Plan and Questions

Above, we discussed how different versions of the app could be bundled together under one product umbrella. The proposed solution for developing these different versions of the app is that each one will be founded, created, and controlled by a different DAO.
<img width="448" src="https://user-images.githubusercontent.com/84213303/150185831-a9b027e8-7369-45be-a709-2873c31d58dd.png">

The specifics for how these versions will be bundled together will be explained in-depth in the client roadmap (#5). The short version is that users will be able to seamlessly choose which version they want to run. Ideally, this will feel like enabling features within the app, while behind the scenes it will be switching app versions.

Users install a specific version of the app controlled by a DAO.
<img width="507" src="https://user-images.githubusercontent.com/84213303/150195828-f66bf9c7-5290-4a53-bbf6-df907ae6f037.png">

The DAO can vote to update the source code for that version.
<img width="586" alt="Screen Shot 2022-01-19 at 11 24 19 AM" src="https://user-images.githubusercontent.com/84213303/150199972-ecd7ac23-033c-4745-9e77-7fc726ceb717.png">

Technically speaking, it's best to think of an app version (i.e. "App 1"  and "App 2") as an **NFT**. Users select which NFT they want to run and the NFT's owner is able to designate the hash of the source code in IPFS/DHT for that app. _This is the decentralized version of how most apps work today. For example, "facebook.com" (the centralized analog to "facebook.eth", which is clearly an NFT) resolves to a particular set of servers that deliver the code for the Facebook web app. Mark Zuckerberg, as the CEO of Facebook, is able to designate which code is executed when people visit Facebook._

At a high-level, this means that DAOs are not an essential part of the Concorde specification. They are a useful abstraction for thinking about how apps ought to be developed, but are technically unnecessary. With that said, in later stages of development, Concorde will lean into the idea of developing apps through DAOs and optimize the system for this use case because providing structure in this way can provide many advantages.

## Phase 0: Pre-Blockchain
**Status:** _In Progress_

The initial version of the app will not use DAOs, smart contracts, or blockchain technology at all. Instead the client will simply retrieve the frontend code based on the hash of the application bundle on IPFS. The default frontend will be determined by the core team and a custom hash will be able to be specified in the application settings.

The purpose of this phase is to get development off the ground. This way frontend developers will be able to begin prototyping their UI and business logic and users will be able to provide feedback based on real usage.

### Milestones
- [ ] Download frontend bundle from IPFS
- [ ] Allow custom hashes so users can specify which frontend they wish to use

### Questions
- **Is there a better solution than IPFS?**

## Phase 1: Unstructured DAOs
**Status:** _Planning_

This phase centers around attaching the app version to an NFT. DAOs can form as a governance mechanism around these NFTs, but the specific DAO implementations will be left up to the app developers. Many small teams of developers may choose a **central point of contact** to manage their NFT, while coordinating their development efforts on Discord, Concorde (dogfooding!), GitHub, or Slack. This will work well because [small groups are effective when centralized](https://github.com/lzrscg/concorde/discussions/3).

App NFTs will resemble [ENS](https://ens.domains) domains. They will have have a unique identifier (most likely just a unique hash, instead of a custom name) and will have customizable metadata. Here is the ENS record for [lzrs.eth](https://app.ens.domains/name/lzrs.eth/details).
<img width="700" src="https://user-images.githubusercontent.com/84213303/150232864-6a8a85e9-3ce3-47af-9633-c26ccda9d6f2.png">

In the case of App NFTs, they will have the following metadata.
| Field | Description |
| --- | --- |
| ID | The unique id of the app |
| Name | The non-unique plaintext name of the app (32 characters) |
| Description | A short description of range app (256 characters) |
| Icon | The hash of a custom icon for this version of the app |
| Source Code | The hash of the source code |
| Owner | The wallet address of the owner |
| Forks | An array of references to other NFTs that are friendly forks |

The `forks` field is of particular importance. This is a list of friendly forks manually set by the owner of the NFT. If users are running a version of the app called "App 1", then the Concorde client will look at App 1's list of friendly forks to determine which alternative versions to suggest to them. It is assumed that this will be a vetted list of good recommendations for App 1's users. Forks will need to negotiate with the owner of App 1 in order to be listed.

Most likely all forks on this list will have compatible data schemas, allowing users to switch back and forth between versions while maintaining full application state. Further discussion of data schemas can be found in the developers tools roadmap (#7).

It is also important to keep in mind that **app versions do not need to be listed** in the `forks` field. There will be another way to switch versions of the app within the client, however, the most convenient method will be from the list of friendly forks.

The `owner` is another important field. In future phases, when apps generate revenue, the owner will be the recipient of all incoming funds.

### Milestones
- [ ] Implement NFT smart contract
- [ ] Resolve app source code and associated information from NFT metadata within the Concorde client
- [ ] Build tooling to help developers create new App NFTs and set metadata

### Questions
- **Which smart contract technology should be used? Ethereum? Solana? Something else? Should it use an L2 or side chain? etc.** The Ethereum ecosystem seems like the best fit since it has the most DeFi applications and integrations (which is important). However, the fees are concerning. In this case, it doesn't seem like many transactions are necessary. But in other cases, such as for microtransactions for messages (see #4), this could be a significant issue. Even getting ETH into an L2 takes a significant amount of gas. It would be ideal not to mix chains—instead falling squarely inside of one L1 ecosystem. With that said, it would be great to see to separate Concorde implementations, one on Ethereum and one on Solana, for example.
- **Is there anything that needs to be done to enable NFTs to be managed by DAOs?** For example, if DAOs want to be able to vote on proposals to update metadata, how can the NFT smart contract be built in order to enable that? What work is necessary to make that happen? In this case, developers would need to implement their own DAOs and there would be little-to-no assumption on Concorde's part about how this would work.

## Phase 2: Structured DAOs
**Status:** _Planning_

This is where things get really interesting. Taking the lessons from unstructured DAOs, Concorde will then begin to engineer DAOs that work enable certain dynamics. One additional assumption here is, at this phase, #4 would be far enough along such that DAOs would have revenue streams based on microtransactions from usage of their app versions.

Here is a list of some potential ideas.

#### Airdrop Forks
In Web3, you can "choose your investors" by airdropping your token to them. DAOs with popular app versions will most likely expect that friendly forks (included on the friendly `fork` list) airdrop a reasonable amount of tokens to them, such as 10-20% of the total supply. This would be a good faith gesture and would allow the original authors of a work to profit off of the success of the derivative (the fork).
<img width="347" src="https://user-images.githubusercontent.com/84213303/150241367-d159848d-56ef-4d18-97bb-b25ebea97e36.png">
This mechanism could even be built into the DAO smart contract and forked app that participates could be automatically added to the friendly fork list. This could either be semi-automated with a quick screening process or default accepted and verified later in case some fork is deemed inappropriate for inclusion.

At the end of the day, the friendly fork list would be at the discretion of the original DAO, but it is also possible that the forking DAO could cut the airdropped tokens out if that original DAO does not hold up their end of the bargain.

#### D2D Investing
One DAO might want to invest in another, especially friendly forks. Using on-chain metrics, it could be possible to implement automated milestone based investments. For example, if a certain number of users switch from using the original to the fork, then the original could have an option to buy X number of the forked tokens at a given price.

Alternatively, this investments could be made in a non-automated way—at the discretion of the community.
#### Mergers and Acquisitions
If another app version is getting used very often—and therefore is earning a lot of revenue—a bigger DAO might decide to "buy the NFT" for that app version. In doing so, they could point it to their app version and, most likely, merge the changes of the code base they are acquiring into their own to keep the acquired user base happy.
<img width="486" src="https://user-images.githubusercontent.com/84213303/150242681-1507f8f8-aa1d-42a3-bb34-6a1580b9c1fb.png">
One could also imagine an arrangement where similar sized DAOs merge.

#### Simultaneous Forks
What if developers of an app version create a modular UI that can enable multiple changes to be automatically merged without conflicts. Friendly forks working within the framework provided by the developers should be able to run their forks alongside other friendly forks also working within that framework.

Here is an example, let's say that there is a polling widget fork and an events widget fork that all fall within a plugin framework developed by a certain app versions. Users should be able to install the polling widget and the event widget forks simultaneously. In this case, the original version, the poll developers, and the events developers should all be compensated. Getting the dynamics right on this will be difficult, but it is technically feasible.

### Milestones
- [ ] Create an official DAO standard and reference smart contract implementation (this will be recommended, but not required)
- [ ] Implement built-in airdrop forking
- [ ] Implement built-in D2D investing
- [ ] Implement built-in M&A
- [ ] Research simultaneous forks
- [ ] Build on existing tooling to help manage all of the above

### Questions
- **How should all of the above be implemented?** More in-depth specifications are needed.
- **Should simultaneous forks be implemented at this phase or the next phase?**
- **What kind of checks and balances are needed to ensure that a majority token holder don't simply transfer App NFTs to themselves, rugging minority token holders?**
- **How should DAO treasuries work?** One idea is that DAO tokens can be burned in exchange for a portion of the treasury. Another idea is that funds go straight to the token holder's wallet.
- **What other similar dynamics could be useful to implement?**

## Phase 3: Integrated Market
**Status:** _Planning_

We will start out without any sort of DAOs. Then we will build NFTs so that DAOs can "own" apps. Then we build standards and tooling so that developers can easily launch DAOs with amazing capabilities. But so far, DAOs have only been a recommendation, and not incorporated directly into Concorde's specification or protocol.

Once all of these dynamics are proven, Concorde can make certain assumptions in order to provide additional functionality. As you may realize, these are getting increasingly speculative. But this is because we are getting further out on the roadmap.
### Integrated Dynamics
The following examples assume there is a Concorde token called $CONCORD. The token for the DAO that owns each respective version of the app will be called $F1, $F2, $F3, etc.

Automated Market Makers provide an on-chain and decentralized mechanism for calculating prices as market caps. As discussed in the tenets, this provides very valuable information. Furthermore, it stands to reason that more popular apps will have higher market caps than less popular ones since the valuation should approximate DFC, which will be in proportion to current and future anticipated usage.

#### Follow the Leader
What will be the default version of Concorde when you open the app? It could be the highest market cap app version, calculated by liquidity pools.
<img width="488" src="https://user-images.githubusercontent.com/84213303/150250508-e3073a9b-b8a0-4461-9c85-3f4ae7083a92.png">
Alternatively, the app could provide a list of potential UIs upon launch ordered by market cap. Until this dynamic is ironed out, it would likely work well to use on-chain usage metrics instead of market cap for rankings. However, this will favor incumbents whereas market cap can price in future potential.

#### Concorde Governance
The UI and business logic of Concorde is super fast moving. This is because it uses forks instead of central planning. But the protocol and client are the infrastructure that enable this functionality—and cannot be forked as easily.

This means it is likely a good idea to have some sort of voting mechanism in which the community has a say in the direction that the protocol and client go. There are a lot of possibilities here. However, we can't have the individual DAOs token holders vote because anyone can create a DAO with a million tokens and then get a million votes. That is, unless DAO tokens can vote weighted by market cap.
<img width="462" src="https://user-images.githubusercontent.com/84213303/150257413-f043c2c1-1bc7-4ea9-8d5f-ce71bde0e71d.png">

#### DAO Marketplace
These DAOs have an opportunity to get access to capital to fund development work. There should be a marketplace that aggregates all of their public work and provides an opportunity for contributors to learn about their roadmap and complete work in exchange for bounties (most likely paid in their own token). These DAOs could gain significant benefit by working according to [GOSPEL](https://www.getrevue.co/profile/lzrs/issues/state-of-gospel-2021-911395).

### Milestones
- [ ] Implement sound, decentralized default app version selection system
- [ ] Implement community governance for the Concorde protocol and client
- [ ] Implement DAO marketplace
- [ ] Investigate the possibility of a Concorde token

### Questions
- **Is a Concorde token necessary for any of this? If so, what other things could it enable?**
- **How would DAOs register to be included in the market cap ranks?** Perhaps if they airdrop some percentage of tokens to the Concorde token holders.
- **What attack vectors are there in liquidity pools? How can these be circumvented?** For example, if there are very few tokens in circulation, it could have an artificially high market cap if the current price is multiplied times the total supply. Market depth and liquidity levels need to be taken into account. I am using the term "market cap" as a proxy for what likely needs to be a more sophisticated metric.
- **How should all of the above be implemented?** Again, more in-depth specifications are needed.
- **What else should be added to this phase?**

**Comment by mateuszsinilo** [comment]
**PHASE 0 QUESTION:** 

In terms of the pre-blockchain phase, I've been talking with a couple of senior devs in a startup, they told me that there's no competitive decentralized p2p storage alternative to IPFS. But I've been thinking if we need IPFS at the early stages? Maybe keeping it around GitHub will be enough for a while to check how it works, and so on, since phase 0 will be so-called "centralized" just to set the platform for people? On the other hand, Github may be good to download the front-end bundle but I don't know if will be enough to maintain a couple of forks (different UIs, features, and so). I think we agree on the bigger vision and specific phase, but we should definitely deep down more in terms of the technical steps of each phase, especially the phases 0 and 1, that will set out our foundations. There are a lot of awesome features that we can apply, and definitely, the token system is something we should consider, especially in terms of subDAOs, which in some sense are token-based communities. But for now, I want to focus on phase 0 and set out a clear system that allows us to push it further ;)
